NTU microphysics scheme

tzuchin12 opened the pull request:
TYPE: new feature

KEYWORDS: NTU microphysics scheme

SOURCE: Tzu-Chin Tsai and Jen-Ping Chen (National Taiwan University)

DESCRIPTION OF CHANGES:
Implementation of the NTU microphysics scheme (mp_physics = 56). 

The NTU (National Taiwan University) scheme applies double moments for the liquid-phase and triple moments for the ice-phase hydrometeors together with the consideration for ice crystal shape and density variations. There are five major features to NTU scheme:  
1. Condensation nuclei (CN) and ice nuclei (IN) are tracked separately for the processes of cloud/rain activation and ice deposition-nucleation using predicted supersaturation;  
2. Applying the triple-moment (the zeroth, second, and third moments) closure method to describe the evolution of ice particle’s spectrum; 
3. Solid-phase hydrometeors’ classification (pristine ice, snow aggregate, rimed ice, and hailstone) is redefined according to their key formation mechanisms; 
4. Ice crystals’ shape and apparent density can evolve gradually according to the growth conditions; 
5. Fall speed of each moment for frozen particles depends on shape and density.  

Recommendations:
Coarse-grid simulations should be okay with the NTU scheme when a time splitting integration is applied; Occasionally, the typical time step of 6dx (grid length of x-direction in kilometers) might be reduced to 4dx or 3dx for resolving the supersaturation explicitly in fine-grid (spacing less than 2 km) simulations. 

Documentation:
For the detailed parameterizations in the NTU scheme, the liquid-based formulae of the bulk conversion rates are adopted from [1], and the technique for aerosol activation to rain/cloud is introduced in [2] and [3]. The triple-moment approach, representation of ice properties, mixed- and solid-based equations/kernels are illustrated in [4] and [5]. In addition to the aerosol background type (CCNTY), some advanced physical options of crystal shape (SHAPE), apparent density (RHOX), fall speed relation (VTX), and bulk method (AFAX) concerning each hydrometeor are described on top of the main module.


LIST OF MODIFIED FILES:
Registry/Registry.EM_COMMON
Registry/Registry.NMM
Registry/registry.var
Registry/registry.wrfplus
dyn_em/solve_em.F
dyn_em/start_em.F
phys/Makefile
phys/module_microphysics_driver.F
phys/module_mp_ntu.F
phys/module_physics_init.F
share/module_bc.F
wrftladj/module_bc_ad.F
wrftladj/module_bc_tl.F
wrftladj/module_microphysics_driver_ad.F
wrftladj/module_microphysics_driver_tl.F
wrftladj/solve_em_ad.F
wrftladj/solve_em_tl.F
run/README.namelist

TESTS CONDUCTED: 
1. jenkins regression tests pass
2. 24-hr summer and winter cases conducted
3. The figure listed below is one real-case example of simulated surface rainfall using different microphysics schemes and different physical treatments within the NTU scheme. 
 

![image](https://user-images.githubusercontent.com/73516426/106420195-13cc6880-6495-11eb-95dd-c68119dbfe35.png)

![image](https://user-images.githubusercontent.com/73516426/106419708-f945bf80-6493-11eb-8d3a-51160222ad61.png)

4. The background aerosols (QDCN, QTCN, and QINI) are needed for the initiation of liquid droplets (activation) and ice particles (nucleation) in the NTU scheme. If we do not specify aerosol boundary conditions, this may cause these inertial aerosols to vanish as a result of advection after a long simulation time. The below figure is one example of a real-case simulation with the default boundary condition (left panels) of QDCN at different time periods. Currently, the zero value (in flow_dep_bdy) and constant of ccn_conc (in flow_dep_bdy_qnn) assigned for inflow in the boundary specified region are not suitable for the vertical damping of aerosols in the NTU scheme. The right panels depict the revised results with the modified flow_dep_bdy_ntu. 

![image](https://user-images.githubusercontent.com/73516426/108478986-21c11c80-72d0-11eb-9e91-ebd0767e3b48.png)


RELEASE NOTE: 

The NTU (National Taiwan University) scheme applies double moments for the liquid-phase and triple moments for the ice-phase hydrometeors together with the consideration for ice crystal shape and density variations. There are five major features to NTU scheme:  condensation nuclei (CN) and ice nuclei (IN) are tracked separately for the processes of cloud/rain activation and ice deposition-nucleation using predicted supersaturation;  applying the triple-moment (the zeroth, second, and third moments) closure method to describe the evolution of ice particle’s spectrum; solid-phase hydrometeors’ classification (pristine ice, snow aggregate, rimed ice, and hailstone) is redefined according to their key formation mechanisms; ice crystals’ shape and apparent density can evolve gradually according to the growth conditions; and fall speed of each moment for frozen particles depends on shape and density.  

For the detailed parameterizations in the NTU scheme, the liquid-based formulae of the bulk conversion rates are adopted from [1], and the technique for aerosol activation to rain/cloud is introduced in [2] and [3]. The triple-moment approach, representation of ice properties, mixed- and solid-based equations/kernels are illustrated in [4] and [5]. In addition to the aerosol background type (CCNTY), some advanced physical options of crystal shape (SHAPE), apparent density (RHOX), fall speed relation (VTX), and bulk method (AFAX) concerning each hydrometeor are described on top of the main module. 

References: 
[1] Chen, J.-P. and S.-T. Liu, 2004: Physically based two-moment bulkwater parameterization for warm-cloud microphysics. Quart. J. Roy. Meteor. Soc., 130, 51–78, doi:10.1256/qj.03.41. 
[2] Cheng, C.-T., W.-C. Wang, and J.-P. Chen, 2007: A modeling study of aerosol impacts on cloud microphysics and radiative properties. Quart. J. Roy. Meteor. Soc., 133, 283–297, doi:10.1002/qj.25. 
[3] Cheng, C.-T., W.-C. Wang, and J.-P. Chen, 2010: Simulation of the effects of increasing cloud condensation nuclei on mixed-phase clouds and precipitation of a front system. Atmos. Res., 96, 461–476, doi:10.1016/j.atmosres.2010.02.005. 
[4] Chen, J.-P. and T.-C. Tsai, 2016: Triple-moment modal parameterization for the adaptive growth habit of pristine ice crystals. J. Atmos. Sci., 73, 2105-2122, doi:10.1175/JAS-D-15-0220.1.
[5] Tsai, T.-C. and J.-P. Chen, 2020: Multi-moment ice bulk microphysics scheme with consideration for particle shape and apparent density. Part I: Methodology and idealized simulation. J. Atmos. Sci., 77, 1821–1850, doi:10.1175/JAS-D-19-0125.1.

weiwangncar commented on Registry/Registry.NMM:
```diff
@@ -871,7 +871,13 @@ state   real    qnwfa          ikjftb  scalar      1         m     i01h023ru=(Up
+state   real    qccn           ijkftb  scalar      1         m     i01h023ru=(UpMass:@ECopy,0.0)d=(DownMass:@ECopy,0.0)f=(BdyMass:@ECopy,0.0)       "QCCN"     "Mass mixing ratio for cloud CN" "kg kg(-1)"
+state   real    qrcn           ijkftb  scalar      1         m     i01h023ru=(UpMass:@ECopy,0.0)d=(DownMass:@ECopy,0.0)f=(BdyMass:@ECopy,0.0)       "QRCN"     "Mass mixing ratio for rain CN" "kg kg(-1)"
+state   real    qnin           ijkftb  scalar      1         m     i01h023ru=(UpMass:@ECopy,0.0)d=(DownMass:@ECopy,0.0)f=(BdyMass:@ECopy,0.0)       "QNIN"     "Potential ice nuclei number mixing ratio" "# kg(-1)"
+#---------------------------------ntu3m--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
@tzuchin12 These are probably needed for the NMM registry.

tzuchin12 replied:
I can not catch the point. They are already listed in Registry.NMM.

weiwangncar replied:
@tzuchin12 I meant new variables qdcn, qtcn, qccn, qrcn and qnin.

tzuchin12 replied:
These five new variables are required to NTU_init through the usage of module_state_description in module_mp_ntu.F.

weiwangncar replied:
@tzuchin12 I don't think these are needed. For example, if you check a variable used by P3 scheme, vmi3d, you will see that this variable, together with some others unique to P3 are not declared in NMM registry.

tzuchin12 replied:
Some modifications were made as suggested.

weiwangncar commented on phys/Makefile:
```diff
@@ -103,6 +103,7 @@ MODULES = \
 	module_mp_wdm6.o \
 	module_mp_wdm7.o \
 	module_mp_cammgmp_driver.o \
+        module_mp_ntu.o \
```
@tzuchin12 Can you line up this?

tzuchin12 replied:
Yes, done.

davegill replied:
@tzuchin12 
This does not seem to be addressed yet. These "line up" requests are tricky. Usually, you need to pull the source code down, and fix it in an editor. The web-based approach has difficulty between distinguishing spaces or tabs.

tzuchin12 replied:
Sorry, it was my misunderstanding. The correction has been made.

weiwangncar commented:
@tzuchin12 Please add your option to run/README.namelist file. Also for the RELEASE NOTE part of the PR, please add a short description of the scheme, in addition to the references.

liujake commented on Registry/registry.var:
```diff
@@ -264,10 +264,9 @@ rconfig   integer   report_end              namelist,wrfvar5  1  10000000 - "rep
 rconfig   integer   tovs_end                namelist,wrfvar5  1  10000000 - "tovs_end"                 ""  ""
 rconfig   logical   gpsref_thinning         namelist,wrfvar5  1  .false.  - "gpsref_thinning"          ""  ""
+rconfig   logical   outer_loop_restart      namelist,wrfvar6  1  .false.  - "outer_loop_restart"       ""  ""
 rconfig   integer   max_ext_its             namelist,wrfvar6  1  1        - "max_ext_its"              ""  ""
 rconfig   integer   ntmax                   namelist,wrfvar6  max_outer_iterations  75      - "ntmax"                    ""  ""
-rconfig   logical   use_inverse_squarerootb namelist,wrfvar6  1  .false.  - "use_inverse_squarerootb"       ""  ""
-rconfig   logical   use_interpolate_cvt     namelist,wrfvar6  1  .false.  - "use_interpolate_cvt"       ""  ""
```
You must use an older version of code here. These three lines of namelist change need to be recovered to the latest 'develop' branch code.

tzuchin12 replied:
Address review comments.

liujake commented:
@tzuchin12 Can you please provide more information/reason for WRFDA-related modifications?, i.e., file # 3, 4, 12-17. Are those modifications just for making WRFPlus still compile with this new MP scheme added? In addition to WTF, did you do any other WRFDA-related compilation and tests? I need to do some WRFDA regression test before approving this PR.

tzuchin12 commented:
> @tzuchin12 Please add your option to run/README.namelist file. Also for the RELEASE NOTE part of the PR, please add a short description of the scheme, in addition to the references.

Modified and updated as suggested.

tzuchin12 commented:
> @tzuchin12 Please add your option to run/README.namelist file. Also for the RELEASE NOTE part of the PR, please add a short description of the scheme, in addition to the references.

@weiwangncar  OK, modified and updated as suggested.

tzuchin12 commented:
> @tzuchin12 Can you please provide more information/reason for WRFDA-related modifications?, i.e., file # 3, 4, 12-17. Are those modifications just for making WRFPlus still compile with this new MP scheme added? 
Yes. 
In addition to WTF, did you do any other WRFDA-related compilation and tests? I need to do some WRFDA regression test before approving this PR.
Only WRFDA-related compilation was checked.

liujake commented:
@tzuchin12 WRFDA regression tests passed. It is good to go for me. Very appreciate for your extra effort to take care of wrfplus-related changes.

davegill commented on Registry/Registry.EM_COMMON:
```diff
@@ -2825,6 +2891,9 @@ package   p3_1category_nc mp_physics==51               -             moist:qv,qc
 package   p3_2category    mp_physics==52               -             moist:qv,qc,qr,qi,qi2;scalar:qnc,qni,qnr,qir,qib,qni2,qir2,qib2;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,vmi3d_2,rhopo3d_2,di3d_2,refl_10cm,th_old,qv_old
 package   morr_tm_aero    mp_physics==40               -             moist:qv,qc,qr,qi,qs,qg;scalar:qnc,qni,qns,qnr,qng;state:rqrcuten,rqscuten,rqicuten,EFCG,EFIG,EFSG,WACT,CCN1_GS,CCN2_GS,CCN3_GS,CCN4_GS,CCN5_GS,CCN6_GS,CCN7_GS,re_cloud,re_ice,re_snow,mskf_refl_10cm
 package   jensen_ishmael  mp_physics==55               -             moist:qv,qc,qr,qi,qi2,qi3;scalar:qnr,qni,qvoli,qaoli,qni2,qvoli2,qaoli2,qni3,qvoli3,qaoli3;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,phii3d,itype,vmi3d_2,rhopo3d_2,di3d_2,phii3d_2,itype_2,vmi3d_3,rhopo3d_3,di3d_3,phii3d_3,itype_3,refl_10cm
+#---------for ntu3m------------------------------------------------------------------------------------------------------------------------------
```
@tzuchin12 
Especially for the package declarations, please remove the 
```
#---------for ntu3m---
```
lines. They are OK for use in the namelist section.

tzuchin12 replied:
Modified as suggested.

davegill commented on Registry/Registry.NMM:
```diff
@@ -1775,6 +1774,9 @@ package   wdm7scheme    mp_physics==26               -             moist:qv,qc,q
 package   p3_1category    mp_physics==50               -             moist:qv
 package   p3_1category_nc mp_physics==51               -             moist:qv
 package   p3_2category    mp_physics==52               -             moist:qv
+#--------------------------------------ntu3m------------------------------------------------------------------------------------------------------------------------------------------
```
@tzuchin12 
Similarly here, you can remove the 
```
#---------for ntu3m---
```
lines for the package statements

tzuchin12 replied:
Modified as suggested.

davegill commented on Registry/registry.var:
```diff
@@ -569,6 +569,9 @@ package   p3_1category    mp_physics==50               -             moist:qv,qc
 package   p3_1category_nc mp_physics==51               -             moist:qv,qc,qr,qi
 package   p3_2category    mp_physics==52               -             moist:qv,qc,qr,qi,qi2
 package   morr_tm_aero    mp_physics==40               -             moist:qv,qc,qr,qi,qs,qg
+#---------------------------------------ntu3m------------------------------------------------------------------------
```
@tzuchin12 
Zap the 
```
#---------for ntu3m---
``` 
here also

tzuchin12 replied:
Modified as suggested.

davegill commented on Registry/registry.wrfplus:
```diff
@@ -820,6 +884,9 @@ package   p3_1category_plus     mp_physics_plus==50     -          g_moist:g_qv,
 package   p3_1category_nc_plus  mp_physics_plus==51     -          g_moist:g_qv,g_qc,g_qr,g_qi;a_moist:a_qv,a_qc,a_qr,a_qi
 package   p3_2category_4dvar    mp_physics_plus==52     -          g_moist:g_qv,g_qc,g_qr,g_qi,g_qi2;a_moist:a_qv,a_qc,a_qr,a_qi,a_qi2
 package   morr_tm_aero_4dvar    mp_physics_plus==40     -          g_moist:g_qv,g_qc,g_qr,g_qi,g_qs,g_qg;a_moist:a_qv,a_qc,a_qr,a_qi,a_qs,a_qg
+#---------------------------------------ntu3m-------------------------------------------------------------------------------------------------------------
```
@tzuchin12 
Last ones here to remove
```
#---------for ntu3m---
```

tzuchin12 replied:
Modified as suggested.

davegill commented on dyn_em/start_em.F:
```diff
@@ -147,6 +148,8 @@ SUBROUTINE start_domain_em ( grid, allowed_to_read &
 #endif
    ALLOCATE(dz8w(IMS:IME,KMS:KME,JMS:JME),STAT=I)  ; dz8w = 0.
    ALLOCATE(z_at_q(IMS:IME,KMS:KME,JMS:JME),STAT=I)  ; z_at_q = 0.
+   ALLOCATE(PRES(IMS:IME,KMS:KME,JMS:JME),STAT=I)  ; PRES = 0.   ! ntu3m
```
@tzuchin12 
Should there be an IF test around this allocate statement? Is it required if the NTU scheme is not run?

tzuchin12 replied:
These two parameters are required for the NTU scheme only.

davegill commented on dyn_em/start_em.F:
```diff
@@ -866,6 +869,8 @@ SUBROUTINE start_domain_em ( grid, allowed_to_read &
       DO k = kts,kte
          DO i = its, min(ite,ide-1)
             z_at_q(i,k,j)=(grid%ph_2(i,k,j)+grid%phb(i,k,j))/g
+            PRES(i,k,j) = grid%p(i,k,j)+grid%pb(i,k,j)              ! ntu3m
```
@tzuchin12 
Here also, and I am assuming below. Are these required if NTU is not the MP chosen?

tzuchin12 replied:
They are required only for the NTU scheme.

davegill commented on phys/module_physics_init.F:
```diff
@@ -49,6 +49,7 @@ SUBROUTINE phy_init ( id, config_flags, DT, restart, zfull, zhalf,     &
                          re_cloud, re_ice, re_snow,              & ! G. Thompson
                          has_reqc, has_reqi, has_reqs,           & ! G. Thompson
 #if ( EM_CORE == 1 )
+                         PRES,RHOA,CCNTY,QDCN,QTCN,QCCN,QRCN,QNIN,& ! for ntu3m
```
@tzuchin12 
Break this into two lines to allow the conventional spacing at the end of the line

tzuchin12 replied:
Modified as suggested.

davegill commented on phys/module_physics_init.F:
```diff
@@ -1517,6 +1525,11 @@ SUBROUTINE phy_init ( id, config_flags, DT, restart, zfull, zhalf,     &
                 adv_moist_cond,                                 &
                 MPDT, DT, DX, DY, LOWLYR,                       &
                 F_ICE_PHY,F_RAIN_PHY,F_RIMEF_PHY,               &
+!-------------------for ntu3m------------------------------------
```
@tzuchin12 
Most developers put a comment on a single line after the "&" character. So remove the 
```
!-------------------for ntu3m--
```
where you do not see others doing this.

tzuchin12 replied:
Modified as suggested.

davegill commented on phys/module_physics_init.F:
```diff
@@ -1517,6 +1525,11 @@ SUBROUTINE phy_init ( id, config_flags, DT, restart, zfull, zhalf,     &
                 F_ICE_PHY,F_RAIN_PHY,F_RIMEF_PHY,               &
+!-------------------for ntu3m------------------------------------
+#if ( EM_CORE == 1 )
+                PRES,RHOA,XLAND,CCNTY,QDCN,QTCN,QCCN,QRCN,QNIN, &
```
@tzuchin12 
Two lines to allow conventional "&" ending

tzuchin12 replied:
Modified as suggested.

davegill commented on share/module_bc.F:
```diff
@@ -2615,6 +2615,113 @@ SUBROUTINE flow_dep_bdy_qnn  (  field,          &
 
    END SUBROUTINE flow_dep_bdy_qnn
 
+!-------------------------for ntu3m-------------------------------------------------
```
@tzuchin12 
Can you briefly describe how this routine is different from the other flow dependent boundary conditions?

davegill replied:
Similarly for the a_ and tladj version

davegill replied:
Sorry, the a_ and the g_versions also. How are these routines different than the existing flow dependent boundary routines?

tzuchin12 replied:
The difference between the flow_dep_bdy_ntu and the other two existing routines is the allocated value to the field. The zero in flow_dep_bdy and one constant of ccn_conc in flow_dep_bdy_qnn are not suitable to be applied in the NTU scheme.

davegill commented on dyn_em/solve_em.F:
```diff
@@ -3754,6 +3765,23 @@ SUBROUTINE solve_em ( grid , config_flags  &
 !       &        , QZH_CURR=scalar(ims,kms,jms,P_QZH), F_QZH=F_QZH          & ! "
       &        , QVOLG_CURR=scalar(ims,kms,jms,P_QVOLG), F_QVOLG=F_QVOLG    & ! for nssl_2mom
       &        , QVOLH_CURR=scalar(ims,kms,jms,P_QVOLH), F_QVOLH=F_QVOLH    & ! for nssl_2mom
+!----------------------------------for ntu3m--------------------------------------------------
```
@tzuchin12 
Typically, developers denote their variables with a comment after the "&" character, instead of 
```
!----------------------------------for ntu3m----
```

tzuchin12 replied:
Modified as suggested.

davegill commented:
@tzuchin12 @weiwangncar @kkeene44 
I have only a few very minor suggestions, regarding indenting, spacing, and comment conventions.

It might be appropriate to show figures of the same case, with a few different (similar) MP schemes. Maybe sensitivities to show the impact of the aerosol type?

Are there restrictions based on time step or grid distance, other suggestions for the users?

Please send a paragraph or two to Kelly for inclusion in the WRF User's Guide. Take a look at https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.2/users_guide_chap5.pdf at the "Summary of Microphysics Options" for a guideline. This is not urgent, but before the end of February.

davegill commented:
Jenkins
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 8dea1410c50054bb0ce6ca548b6e721f5ee2ed3b, requested by: tzuchin12 for PR: https://github.com/wrf-model/WRF/pull/1364. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 163           161        0
    Number of Comparisons  : 103           102        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

davegill commented on wrftladj/solve_em_ad.F:
```diff
@@ -3015,8 +3014,8 @@ SUBROUTINE solve_em_ad ( grid , config_flags  &
-                                   grid%j_end(ij),k_start,k_end)
-!             IF(is .ne. P_QNN)THEN
+                                   grid%j_end(ij),k_start,k_end)        ! for ntu3m
+!ntu3m             IF(is .ne. P_QNN)THEN
```
@tzuchin12 
It looks like this code is purposefully commented out. Place your "!ntu3m" identifier AFTER the Fortran statement. Also at line 3039.

tzuchin12 replied:
Corrections have been made.

davegill commented on wrftladj/solve_em_tl.F:
```diff
@@ -2782,8 +2781,8 @@ SUBROUTINE solve_em_tl ( grid , config_flags  &
-                                     k_start,k_end)
-!             IF(is .ne. P_QNN)THEN
+                                     k_start,k_end)                      ! for ntu3m
+!ntu3m             IF(is .ne. P_QNN)THEN
```
@tzuchin12 
Another example to place the comment AFTER the Fortran statement (even though the Fortran statement is commented out).

tzuchin12 replied:
Done.

davegill commented on wrftladj/solve_em_tl.F:
```diff
@@ -2803,8 +2802,7 @@ SUBROUTINE solve_em_tl ( grid , config_flags  &
 !                                  grid%j_start(ij), grid%j_end(ij),  &
-!                                  k_start, k_end                    )
-!----------------------------for ntu3m---------------------------------
+!ntu3m                                  k_start, k_end                    )
```
... and here, too

tzuchin12 replied:
Done.

davegill commented:
@tzuchin12 
A GREAT JOB putting the comments into a standard form.

davegill commented:
@tzuchin12 @kkeene44 
Here are outstanding issues to address. They were mentioned above, but you have handled some of the original list, and these are remaining:

It is appropriate to show figures of the same case, with a few different (similar) MP schemes. In your case, you could include sensitivities to show the impact of the aerosol options.

Are there restrictions based on time step or grid distance, other suggestions for the users? Are there cases that this works particularly well?

Please send a paragraph or two to Kelly for inclusion in the WRF User's Guide. Take a look at https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.2/users_guide_chap5.pdf at the "Summary of Microphysics Options" for a guideline. This is not urgent, but before the end of February.

weiwangncar commented:
@tzuchin12 @davegill Regarding the text in the UG, Tzu-Chin will provide me the paragraph and I will pass it to Kelly. We have also discussed preparing some simulation results to present here.

tzuchin12 commented:
@davegill @weiwangncar  The introduction to the NTU scheme in PR is updated, as well as the paragraph. Please let me know if anything I can do.

davegill commented:
@tzuchin12 @weiwangncar @dudhia
This PR is shaping up very nicely. I think that all of the specific questions have been addressed. From a software perspective, this PR is ready.

smileMchen commented:
@davegill @weiwangncar @dudhia @tzuchin12 
All,
Before you approve this PR and merge to repository, I would like to run a few real-data cases just to make sure it works fine. I will keep you updated after I am done. Thanks.

dudhia commented on share/module_bc.F:
```diff
@@ -2615,6 +2615,113 @@ SUBROUTINE flow_dep_bdy_qnn  (  field,          &
+                               ide,jds,jde,kds,kde,ims,ime,jms,jme,kms,&
+                               kme,ips,ipe,jps,jpe,kps,kpe,its,ite,jts,&
+                               jte,kts,kte)
+!-------------------------------------------------------------------------------
```
I would like to know why this new boundary condition is needed and especially why changes are needed for periodic conditions. The Registry has the PERIOD halos that may address these.

tzuchin12 replied:
The background aerosols (QDCN, QTCN, and QINI) are needed for the initiation of liquid droplets (activation) and ice particles (nucleation) in the NTU scheme. If we do not specify aerosol boundary conditions, this may cause these inertial aerosols to vanish as a result of advection after a long simulation time. The below figure is one example of a real-case simulation with the default boundary condition (left panels) of QDCN at different time periods. Currently, the zero value (in flow_dep_bdy) and constant of ccn_conc (in flow_dep_bdy_qnn) assigned for inflow in the boundary specified region are not suitable for the vertical damping of aerosols in the NTU scheme. The right panels depict the revised results with the modified flow_dep_bdy_ntu. It will be appreciated if you have any suggestions. 

![image](https://user-images.githubusercontent.com/73516426/108478986-21c11c80-72d0-11eb-9e91-ebd0767e3b48.png)

dudhia commented:
Do the background values have a profile or fixed value?
flow_dep_bdy_qnn allows a fixed value input.

On Fri, Feb 19, 2021 at 1:59 AM tzuchin12 <notifications@github.com> wrote:

> *@tzuchin12* commented on this pull request.
> ------------------------------
>
> In share/module_bc.F
> <https://github.com/wrf-model/WRF/pull/1364#discussion_r579026426>:
>
> > @@ -2615,6 +2615,113 @@ SUBROUTINE flow_dep_bdy_qnn  (  field,          &
>
>     END SUBROUTINE flow_dep_bdy_qnn
>
> +!-------------------------for ntu3m-------------------------------------------------
> +   SUBROUTINE flow_dep_bdy_ntu(field,u,v,config_flags,spec_zone,ids,   &
> +                               ide,jds,jde,kds,kde,ims,ime,jms,jme,kms,&
> +                               kme,ips,ipe,jps,jpe,kps,kpe,its,ite,jts,&
> +                               jte,kts,kte)
> +!-------------------------------------------------------------------------------
>
> The background aerosols (QDCN, QTCN, and QINI) are needed for the
> initiation of liquid droplets (activation) and ice particles (nucleation)
> in the NTU scheme. If we do not specify aerosol boundary conditions, this
> may cause these inertial aerosols to vanish as a result of advection after
> a long simulation time. The below figure is one example of a real-case
> simulation with the default boundary condition (left panels) of QDCN at
> different time periods. Currently, the zero value (in flow_dep_bdy) and
> constant of ccn_conc (in flow_dep_bdy_qnn) assigned for inflow in the
> boundary specified region are not suitable for the vertical damping of
> aerosols in the NTU scheme. The right panels depict the revised results
> with the modified flow_dep_bdy_ntu. It will be appreciated if you have any
> suggestions.
>
> [image: image]
> <https://user-images.githubusercontent.com/73516426/108478986-21c11c80-72d0-11eb-9e91-ebd0767e3b48.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1364#discussion_r579026426>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77D3HEVMYJS2Y5WOUILS7YSAZANCNFSM4WB7KV2A>
> .
>

tzuchin12 commented:
> Do the background values have a profile or fixed value? flow_dep_bdy_qnn allows a fixed value input.
The different initial values of background aerosols are derived from ntu_init (not fixed values, and there are four typical CCN types to choose).  Also, the vertical distribution is decreased exponentially with a scale height (3.57 km for continental and 0.8 km for marine type) except for the lowest three-sigma levels or below 850 hPa. So, it is not suitable to simply assign a profile or fixed value in the NTU scheme using flow_dep_bdy_qnn.

dudhia commented:
OK, this keeps the inflow value constant at the initial value.
We can rename the routine more generally like flow_dep_bdy_fixed_inflow

On Fri, Feb 19, 2021 at 10:53 AM tzuchin12 <notifications@github.com> wrote:

> Do the background values have a profile or fixed value? flow_dep_bdy_qnn
> allows a fixed value input.
> The different initial values of background aerosols are derived from
> ntu_init (not fixed values, and there are four typical CCN types to
> choose). Also, the vertical distribution is decreased exponentially with a
> scale height (3.57 km for continental and 0.8 km for marine type) except
> for the lowest three-sigma levels or below 850 hPa. So, it is not suitable
> to simply assign a profile or fixed value in the NTU scheme using
> flow_dep_bdy_qnn.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-782235423>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77CAYVLD7W2ZDDF4EBDS72QP3ANCNFSM4WB7KV2A>
> .
>

tzuchin12 commented:
> OK, this keeps the inflow value constant at the initial value. We can rename the routine more generally like flow_dep_bdy_fixed_inflow
R: Thanks for the suggestion. The name of flow_dep_bdy_ntu has been changed to flow_dep_bdy_fixed_inflow over the related modules.

dudhia reviewed (approved):
also approve

twjuliano commented:
@tzuchin12 @weiwangncar @davegill @dudhia
I think that there may have been a bug introduced by this PR in start_em.F, specifically on these lines (872-873):

`PRES(i,k,j) = grid%p(i,k,j)+grid%pb(i,k,j)              ! ntu3m`
`RHOA(i,k,j) = 1./grid%alt(i,k,j)*(1.+moist(i,k,j,P_QV)) ! ntu3m`

The loop goes from k=kts,kte where kte=kpe. Each of the variables on the RHS is defined at mass points, hence there should probably be a separate loop for these variables defined from k=kts,kte-1. The code fails on the second line above for an idealized simulation, as it tries to divide by zero.

weiwangncar commented:
@twjuliano Thanks for pointing this out. I agree another do loop may be needed here.

dudhia commented:
Yes, I think that loop should go to kte-1. Good catch.

On Tue, Mar 16, 2021 at 11:38 AM Timothy Juliano ***@***.***>
wrote:

> @tzuchin12 <https://github.com/tzuchin12> @weiwangncar
> <https://github.com/weiwangncar> @davegill <https://github.com/davegill>
> @dudhia <https://github.com/dudhia>
> I think that there may have been a bug introduced by this PR in
> start_em.F, specifically on these lines (872-873):
>
> PRES(i,k,j) = grid%p(i,k,j)+grid%pb(i,k,j) ! ntu3m
> RHOA(i,k,j) = 1./grid%alt(i,k,j)*(1.+moist(i,k,j,P_QV)) ! ntu3m
>
> The loop goes from k=kts,kte where kte=kpe. Each of the variables on the
> RHS is defined at mass points, hence there should probably be a separate
> loop for these variables defined from k=kts,kte-1. The code fails on the
> second line above for an idealized simulation, as it tries to divide by
> zero.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-800469039>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77H6375Z4FW2ANIVJ5DTD6JQDANCNFSM4WB7KV2A>
> .
>

dudhia commented:
z_at_q seems to be for mass levels so that formula using ph is off by half
a level.

On Tue, Mar 16, 2021 at 12:12 PM Jimy Dudhia ***@***.***> wrote:

> Yes, I think that loop should go to kte-1. Good catch.
>
> On Tue, Mar 16, 2021 at 11:38 AM Timothy Juliano ***@***.***>
> wrote:
>
>> @tzuchin12 <https://github.com/tzuchin12> @weiwangncar
>> <https://github.com/weiwangncar> @davegill <https://github.com/davegill>
>> @dudhia <https://github.com/dudhia>
>> I think that there may have been a bug introduced by this PR in
>> start_em.F, specifically on these lines (872-873):
>>
>> PRES(i,k,j) = grid%p(i,k,j)+grid%pb(i,k,j) ! ntu3m
>> RHOA(i,k,j) = 1./grid%alt(i,k,j)*(1.+moist(i,k,j,P_QV)) ! ntu3m
>>
>> The loop goes from k=kts,kte where kte=kpe. Each of the variables on the
>> RHS is defined at mass points, hence there should probably be a separate
>> loop for these variables defined from k=kts,kte-1. The code fails on the
>> second line above for an idealized simulation, as it tries to divide by
>> zero.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-800469039>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEIZ77H6375Z4FW2ANIVJ5DTD6JQDANCNFSM4WB7KV2A>
>> .
>>
>

dudhia commented:
z_at_q was added for Thompson aerosols and is not part of the NTU change.

On Tue, Mar 16, 2021 at 12:15 PM Jimy Dudhia ***@***.***> wrote:

> z_at_q seems to be for mass levels so that formula using ph is off by half
> a level.
>
> On Tue, Mar 16, 2021 at 12:12 PM Jimy Dudhia ***@***.***> wrote:
>
>> Yes, I think that loop should go to kte-1. Good catch.
>>
>> On Tue, Mar 16, 2021 at 11:38 AM Timothy Juliano <
>> ***@***.***> wrote:
>>
>>> @tzuchin12 <https://github.com/tzuchin12> @weiwangncar
>>> <https://github.com/weiwangncar> @davegill <https://github.com/davegill>
>>> @dudhia <https://github.com/dudhia>
>>> I think that there may have been a bug introduced by this PR in
>>> start_em.F, specifically on these lines (872-873):
>>>
>>> PRES(i,k,j) = grid%p(i,k,j)+grid%pb(i,k,j) ! ntu3m
>>> RHOA(i,k,j) = 1./grid%alt(i,k,j)*(1.+moist(i,k,j,P_QV)) ! ntu3m
>>>
>>> The loop goes from k=kts,kte where kte=kpe. Each of the variables on the
>>> RHS is defined at mass points, hence there should probably be a separate
>>> loop for these variables defined from k=kts,kte-1. The code fails on the
>>> second line above for an idealized simulation, as it tries to divide by
>>> zero.
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-800469039>, or
>>> unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AEIZ77H6375Z4FW2ANIVJ5DTD6JQDANCNFSM4WB7KV2A>
>>> .
>>>
>>

weiwangncar commented:
@dudhia Since z_at_q uses ph, it should be just height at full levels, whether it is intended or not.

dudhia commented:
I don't think it is intended. There is already z_at_w for full levels.

On Tue, Mar 16, 2021 at 12:30 PM weiwangncar ***@***.***>
wrote:

> @dudhia <https://github.com/dudhia> Since z_at_q uses ph, it should be
> just height at full levels, whether it is intended or not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-800506493>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77D36YG47C4KOVMEHZ3TD6PUVANCNFSM4WB7KV2A>
> .
>

weiwangncar commented:
@dudhia Then that z_at_q calculation needs to be changed, and it may affect results from Thompson scheme. I'm ok with the change, but do we need Greg's blessing?

dudhia commented:
I think we need to test the effect. It seems to be only used to initialize
the surface aerosol source.

On Tue, Mar 16, 2021 at 1:27 PM weiwangncar ***@***.***>
wrote:

> @dudhia <https://github.com/dudhia> Then that z_at_q calculation needs to
> be changed, and it may affect results from Thompson scheme. I'm ok with the
> change, but do we need Greg's blessing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-800543021>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77AHVUQTAYZYFT4TCWDTD6WI5ANCNFSM4WB7KV2A>
> .
>

dudhia commented:
@Greg Thompson ***@***.***> Check which hgt you need in
thompson_init for aerosols.
Currently these are full level values but the name is z_at_q in start_em
which
seems to be to distinguish from z_at_w that is the full-level one. Should
we just
use z_at_w? From thompson_init, it seems the full level ones make sense.

On Tue, Mar 16, 2021 at 1:36 PM Jimy Dudhia ***@***.***> wrote:

> I think we need to test the effect. It seems to be only used to initialize
> the surface aerosol source.
>
> On Tue, Mar 16, 2021 at 1:27 PM weiwangncar ***@***.***>
> wrote:
>
>> @dudhia <https://github.com/dudhia> Then that z_at_q calculation needs
>> to be changed, and it may affect results from Thompson scheme. I'm ok with
>> the change, but do we need Greg's blessing?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-800543021>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEIZ77AHVUQTAYZYFT4TCWDTD6WI5ANCNFSM4WB7KV2A>
>> .
>>
>

gthompsnWRF commented:
I am pretty sure that I need `z_at_w` because I need the "thickness" of the first model layer that represents the depth of atmosphere where the thermodynamic variables are defined.  That way, I compute an "airmass" variable that is used in the aerosol-aware surface emissions scaling.

dudhia commented:
Yes, the name z_at_q is confusing. We could leave it for now, however.

On Mon, Mar 22, 2021 at 11:52 AM gthompsnWRF ***@***.***>
wrote:

> I am pretty sure that I need z_at_w because I need the "thickness" of the
> first model layer that represents the depth of atmosphere where the
> thermodynamic variables are defined. That way, I compute an "airmass"
> variable that is used in the aerosol-aware surface emissions scaling.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1364#issuecomment-804269285>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77B5ELRLNIPTQ6IWPHTTE57WHANCNFSM4WB7KV2A>
> .
>
